fms/modules/attention.py

```diff
@@ -83,28 +87,23 @@ def __init__(
 
     def reset_params(self, gain=1):
         # Ensure softmax inputs are standard normal
-        for layer in ["query", "key"]:
-            nn.init.trunc_normal_(
-                getattr(self, layer).weight, mean=0.0, std=self.emb_dim**-0.5
-            )
+        self.qkv_fused.weight.data[
+            : self.emb_kq_per_head * (self.nheads + self.kvheads)
+        ].normal_(0, self.emb_dim**-0.5)
         # Ensure projection layers have same scale (for normalized-step dataloaders like
         # AdamW / Sophia), and maintain input norm up to attention remix, in expectation
-        for layer in ["value", "dense"]:
-            nn.init.trunc_normal_(
-                getattr(self, layer).weight,
-                mean=0.0,
-                std=(gain / (self.emb_dim * self.nheads * self.emb_v_per_head) ** 0.5)
-                ** 0.5,
-            )  # Using explicit terms instead of numel to account for eventual MQA addition
+        self.qkv_fused.weight.data[
+            self.emb_kq_per_head * (self.nheads + self.kvheads) :
+        ].normal_(
+            0, (gain / (self.emb_dim * self.nheads * self.emb_v_per_head) ** 0.5) ** 0.5
+        )
         if self.use_bias:
-            for layer in ["query", "key", "value", "dense"]:
+            for layer in ["qkv_fused", "dense"]:
                 getattr(self, layer).bias.data.zero_()
 
     def forward(
         self,
-        q,
-        k,
-        v,
+        qkv,
```

Comment:
I think we want to continue to offer the option to have a kq length taht differs from the v length. In some implementations of MHA I've seen them accept q,k,v and then check if the 3 tensors are identical to determine whether to treat them as separate or just pick one.

Reply:
e.g. in this impl there's lots of `if query is key` handling:

https://pytorch.org/docs/stable/_modules/torch/nn/modules/activation.html#MultiheadAttention

or in xformers there's code where k/v are assumed to be q if unspecified:
https://github.com/facebookresearch/xformers/blob/main/xformers/components/multi_head_dispatch.py#L165

Reply:
I agree, will switch this to taking in q, k, v separately

Reply:
qkv is now split into q, k, v params in forward. k and v are optional, if None will set to q, otherwise will cat all q, k, and v

Reply:
We may need to think about this for cross attention as well (though not sure as part of this PR)